File: android/app/src/test/java/jp/touya/app/domain/PresenceTest.kt
package jp.touya.app.domain

import jp.touya.app.data.Absences
import jp.touya.app.data.CharacterPresence
import jp.touya.app.data.Streaks
import jp.touya.app.data.TodayLine
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class PresenceTest {
    private val nightSaturday = Clock(
        day = "2026-09-12",
        hour = 22,
        month = 9,
        weekday = "sat",
        part = "night",
    )

    private val presence = CharacterPresence(
        today = listOf(
            TodayLine("いつもの席。", whenPart = "any"),
            TodayLine("土曜の夜、席はあけてある。", whenPart = "night", weekday = "sat"),
            TodayLine("朝の光。", whenPart = "dawn"),
        ),
        hooks = listOf("また明日、テラスで。"),
        absences = Absences(
            short = listOf("昨日は来なかったね。"),
            few = listOf("数日、空いてた。今日は隣でいい。"),
            week = listOf("しばらく、空いてた。"),
        ),
        streaks = Streaks(
            three = listOf("三日続いてる。"),
            seven = listOf("一週間、来てる。"),
        ),
        suggestionsByStage = mapOf(
            "regular" to listOf("いつもの席でいい？"),
        ),
    )

    @Test
    fun todayLinePrefersNightAndSaturday() {
        val text = pickTodayLine(presence, nightSaturday, "regular", "hiyori:2026-09-12")
        assertEquals("土曜の夜、席はあけてある。", text)
    }

    @Test
    fun sameSeedReturnsSameLine() {
        val a = pickTodayLine(presence, nightSaturday, "familiar", "rione:2026-09-12")
        val b = pickTodayLine(presence, nightSaturday, "familiar", "rione:2026-09-12")
        assertEquals(a, b)
    }

    @Test
    fun openingNoticesThreeDayAbsenceWithoutBlame() {
        val opening = composeOpening(
            id = "shiraishi",
            greeting = "……来たの。",
            welcomeBack = "また来た。",
            presence = presence,
            clock = nightSaturday,
            stage = "familiar",
            daysAway = 4,
            streak = 1,
            firstVisit = false,
        )
        assertEquals("absence", opening.kind)
        assertFalse(Regex("怒|許さ|裏切り").containsMatchIn(opening.text))
        assertTrue(opening.text.contains("数日、空いてた"))
    }

    @Test
    fun hookRecoveryComesBeforeToday() {
        val opening = composeOpening(
            id = "clara",
            greeting = "Bonsoir.",
            welcomeBack = "おかえりなさい。",
            presence = presence,
            clock = nightSaturday,
            stage = "regular",
            daysAway = 1,
            streak = 2,
            hook = "また明日、テラスで。",
            firstVisit = false,
        )
        assertEquals("hook", opening.kind)
        assertTrue(opening.text.contains("テラス"))
    }

    @Test
    fun suggestionsPreferStage() {
        val hints = suggestionsFor(presence, "regular", listOf("fallback"))
        assertEquals(listOf("いつもの席でいい？"), hints)
    }
}

File: src/lib/presence.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import type { Clock } from "./clock";
import { composeOpening, pickTodayLine } from "./presence";
import type { CharacterPresence } from "./presence-types";

const nightSaturday: Clock = {
  day: "2026-09-12",
  hour: 22,
  month: 9,
  weekday: "sat",
  part: "night",
};

function load(id: string): CharacterPresence {
  return JSON.parse(readFileSync(join(process.cwd(), "shared/presence", `${id}.json`), "utf8"));
}

test("today line prefers night-and-saturday over a generic line", () => {
  const presence = load("hiyori");
  const text = pickTodayLine(presence, nightSaturday, "regular", "hiyori:2026-09-12");
  assert.ok(text);
  assert.match(text, /土曜|夜|席/);
});

test("same seed returns the same today line", () => {
  const presence = load("rione");
  const a = pickTodayLine(presence, nightSaturday, "familiar", "rione:2026-09-12");
  const b = pickTodayLine(presence, nightSaturday, "familiar", "rione:2026-09-12");
  assert.equal(a, b);
});

test("opening notices a three-day absence without blaming", () => {
  const opening = composeOpening({
    id: "shiraishi",
    greeting: "……来たの。",
    welcomeBack: "また来た。",
    presence: load("shiraishi"),
    clock: nightSaturday,
    stage: "familiar",
    daysAway: 4,
    streak: 1,
    firstVisit: false,
  });
  assert.equal(opening.kind, "absence");
  assert.doesNotMatch(opening.text, /怒|許さ|裏切り/);
});

test("hook recovery comes before today's weather", () => {
  const opening = composeOpening({
    id: "clara",
    greeting: "Bonsoir.",
    welcomeBack: "おかえりなさい。",
    presence: load("clara"),
    clock: nightSaturday,
    stage: "regular",
    daysAway: 1,
    streak: 2,
    hook: "また明日、テラスで。",
    firstVisit: false,
  });
  assert.equal(opening.kind, "hook");
  assert.match(opening.text, /テラス/);
});
